src/translator/cache.h

```diff
@@ -0,0 +1,196 @@
+  // point something that is alive for the duration, the contents of this are memcpy'd into L4s internal storage.
+  //
+  // To hide the cryptic names, this class uses friendly "KeyBytes" and "ValueBytes", which is what they essentially
+  // are.
+  using KeyBytes = L4::IWritableHashTable::Key;
+  using ValueBytes = L4::IWritableHashTable::Value;
+
+  /// Construct a ThreadSafeL4Cache from configuration passed via key values in an Options object:
+  ///
+  /// @param [in] options: Options object which is used to configure the cache. See command-line parser for
+  /// documentation (`marian::bergamot::createConfigParser()`)
+  ThreadSafeL4Cache(Ptr<Options> options);
+
+  /// Fetches a record from cache, storing it in processedRequestSentence if found. Calls to fetch are thread-safe and
+  /// lock-free.
+  ///
+  /// @param [in]  words: query marian:Words for which translation results are looked up in cache.
+  /// @param [out] processedRequestSentence: stores cached results for use outside if found.
+  ///
+  /// @returns true if query found in cache false otherwise.
+  bool fetch(const marian::Words &words, ProcessedRequestSentence &processedRequestSentence);
+
+  /// Inserts a new record into cache. Thread-safe. Modifies the structure so takes locks, configure sharding to
+  /// configure how fine-grained the locks are to reduce contention.
+  ///
+  /// @param [in] words: marian::Words processed from a sentence
+  /// @param [in] processedRequestSentence: minimum translated information corresponding to words
+  ///
+  void insert(const marian::Words &words, const ProcessedRequestSentence &processedRequestSentence);
+
+  CacheStats stats() const;
+
+ private:
+  /// cacheConfig {sizeInBytes, recordTimeToLive, removeExpired} determins the upper limit on cache storage, time for
+  /// a record to live and whether or not to evict expired records.
+  L4::HashTableConfig::Cache cacheConfig_;
+
+  /// epochManagerConfig_{epochQueueSize, epochProcessingInterval, numActionQueues}:
+  /// - epochQueueSize (How many actions to keep in a Queue)
+  /// - epochProcessingInterval (Reclaimation runs in a background thread. The interval at which this thread is run).
+  /// - numActionQueues (Possibly to increase throughput of registering actions by a finer-granularity of locking).
+  /// .. and "action" is an std::function<void(void)>..., which is run by the background thread.
+  L4::EpochManagerConfig epochManagerConfig_;
+
+  /// L4 Service. There's a string key based registration through service_ and accesses from multiple processes to get
+  /// the same map.
+  L4::LocalMemory::HashTableService service_;
+
+  /// An L4 Context, does magic with epochs everytime accessed;  "Once a context is retrieved, the operations such as
+  /// operator[] on the context and Get() are lock-free."
+  L4::LocalMemory::Context context_;
+
+  /// context_[hashTableIndex_] gives the hashmap for Get(...) or Add(...) operations
+  size_t hashTableIndex_;
```

Comment:
Do we have a use case where hashTableIndex != 1

Reply:
I hope they used 0-based indexing. `hashTableIndex_ != 0` in this case? To answer your query, we probably don't. There's only 1 Cache we use which we have an adapter wrapping around. L4 however in the example provides only doc on how to operate it with the service, string and indices.

Reply:
Yes, I meant 0.  We should hardcode that to 0 and remove the unnecessary array lookups (for the clarity of the person reading the code)

Reply:
Hmmm, do we want to do this? We're taking up the responsibility of manually verifying things here - why not simply keep L4 as the single-source of truth?

Reply:
All i am saying is `const size_t constexpr hashTableIndex_ = 1;`. That way when you open any IDE you would hover it and find "oh, we just have one hash table, ok".

Reply:
> `const size_t constexpr hashTableIndex_ = 1;` 

You mean zero again, I take. That this human error happens again is reason enough for me not to change this. xD

Reply:
Yes, because passing that parameter from somewhere else is so much less error prone. Logic.